sample/signalr-server-selfhost/Connections/RawConnection.cs

```diff
@@ -0,0 +1,170 @@
+            return Connection.Broadcast(DateTime.Now + ": from " + connectionId + " OnDisconnected, stopCalled: " + stopCalled);
+        }
+
+        protected override Task OnReceived(IRequest request, string connectionId, string data)
+        {
+
+            var message = JsonConvert.DeserializeObject<Message>(data);
+
+            switch (message.Type)
+            {
+                case MessageType.Broadcast:
+                    Connection.Broadcast("from " + connectionId + ": Broadcast: " + message.Value);
+                    break;
+                case MessageType.BroadcastExceptMe:
+                    Connection.Broadcast("from " + connectionId + ": Broadcast except me: " + message.Value, connectionId);
+                    break;
+                case MessageType.Send:
+                    Connection.Send(connectionId, "from " + connectionId + ": Send to Caller: " + message.Value);
+                    break;
+                case MessageType.Specified:
+                    Connection.Send(message.Connection, "from " + connectionId + ": Send to Client: " + message.Value);
+                    break;
+                case MessageType.Clients:
+                    Connection.Send(message.Connections, "from " + connectionId + ": Send to Clients: " + message.Value);
+                    break;
+                case MessageType.Join:
+                    string name = message.Value;
+                    _clients[connectionId] = name;
+                    _users[name] = connectionId;
+                    Connection.Send(connectionId, new
+                    {
+                        from = GetUser(connectionId),
+                        type = MessageType.Join.ToString(),
+                        data = message.Value
+                    });
+                    break;
+                case MessageType.PrivateMessage:
+                    var parts = message.Value.Split('|');
+                    string user = parts[0];
+                    string msg = parts[1];
+                    string id = GetClient(user);
+                    Connection.Send(id, new
+                    {
+                        from = GetUser(connectionId),
+                        type = MessageType.PrivateMessage.ToString(),
+                        data = msg
+                    });
+                    break;
+                case MessageType.AddToGroup:
+                    Groups.Add(connectionId, message.Value).Wait();
+                    Groups.Add(connectionId, message.Value).Wait();
+                    Groups.Add(connectionId, message.Value).Wait();
+                    Connection.Send(connectionId, "[" + DateTime.Now.ToShortTimeString() + "]: Succeeded at joinGroup: " + connectionId + " joined " + message.Value);
+                    break;
+                case MessageType.RemoveFromGroup:
+                    Groups.Remove(connectionId, message.Value).Wait();
+                    Groups.Remove(connectionId, message.Value).Wait();
```

Comment:
apart from `.Wait()` - why would removing the connection from the Group twice make a difference **for the client**?